stdlib/LinearAlgebra/src/special.jl

```diff
@@ -148,3 +148,8 @@ function fill!(A::Union{Diagonal,Bidiagonal,Tridiagonal,SymTridiagonal}, x)
     throw(ArgumentError("array of type $(typeof(A)) and size $(size(A)) can
     not be filled with $x, since some of its entries are constrained."))
 end
+
+one(A::Diagonal{T}) where T = Diagonal(fill!(similar(A.diag), one(T)))
+one(A::Bidiagonal{T}) where T = Bidiagonal(fill!(similar(A.dv), one(T)), zero(A.ev), A.uplo)
+one(A::Tridiagonal{T}) where T = Tridiagonal(zero(A.du), fill!(similar(A.d), one(T)), zero(A.dl))
+one(A::SymTridiagonal{T}) where T = SymTridiagonal(fill!(similar(A.dv), one(T)), zero(A.ev))
```

Comment:
`similar(foo)` in all of these cases is wrong.   It should be `similar(foo, typeof(one(T)))`

The issue is that, if `T` is a dimensionful type, `one` returns a dimensionless type.

Reply:
You could also use `fill(one(T), size(foo))`

Reply:
`zero(A.dl)` etcetera is wrong for a related reason, because `zero` returns a dimensionful value.   You should instead use `fill(zero(one(T)), size(A.dl))` or similar.

Reply:
> You could also use fill(one(T), size(foo)), which might be simpler.

No, this would always return `Vector`.

Reply:
Would be good to test the dimensionful case … see the tests with `Furlongs` in the `test/triangular.jl` file.

Reply:
Here we would like to have back `one(x)`, which would also work with static vectors.

Reply:
Here we would like to have back `ones(x)` which would also work with static arrys.

Reply:
> No, this would always return `Vector`.

No, that's not true.  `size` returns a tuple, and `fill` for a tuple argument returns an array of the corresponding dimensionality.

Reply:
@fredrikekre is correct about (not) using `fill(one(T), size(foo))`. This would not preserve the underlying container type (as an example: sparsevector).

As for the dimensionless type, using similar should an abstractvector with the given eltype, which then promotes the result of `one` to the appropriate dimensionful type.

For example:

```
julia> x = Furlong(3)
Furlong{1,Int64}(3)

julia> D = Diagonal([x, x, x])
3×3 Diagonal{Furlong{1,Int64},Array{Furlong{1,Int64},1}}:
 Furlong{1,Int64}(3)           ⋅                    ⋅
          ⋅           Furlong{1,Int64}(3)           ⋅
          ⋅                    ⋅           Furlong{1,Int64}(3)

julia> one(D)
3×3 Diagonal{Furlong{1,Int64},Array{Furlong{1,Int64},1}}:
 Furlong{1,Int64}(1)           ⋅                    ⋅
          ⋅           Furlong{1,Int64}(1)           ⋅
          ⋅                    ⋅           Furlong{1,Int64}(1)

julia> y = Furlong{2}(3)
Furlong{2,Int64}(3)

julia> E = Diagonal([y, y, y])
3×3 Diagonal{Furlong{2,Int64},Array{Furlong{2,Int64},1}}:
 Furlong{2,Int64}(3)           ⋅                    ⋅
          ⋅           Furlong{2,Int64}(3)           ⋅
          ⋅                    ⋅           Furlong{2,Int64}(3)

julia> one(E)
3×3 Diagonal{Furlong{2,Int64},Array{Furlong{2,Int64},1}}:
 Furlong{2,Int64}(1)           ⋅                    ⋅
          ⋅           Furlong{2,Int64}(1)           ⋅
          ⋅                    ⋅           Furlong{2,Int64}(1)

```

Is this not the behavior that we want?

Reply:
`one` should return a multiplicative identity, so it should be unitless, e.g.
```
julia> x = Furlong(3)
Furlong{1,Int64}(3)

julia> one(x)
1

julia> x * one(x) == x
true
```
and thus,
```
julia> D = Diagonal([x, x, x])
3×3 Diagonal{Furlong{1,Int64},Array{Furlong{1,Int64},1}}:
 Furlong{1,Int64}(3)           ⋅                    ⋅         
          ⋅           Furlong{1,Int64}(3)           ⋅         
          ⋅                    ⋅           Furlong{1,Int64}(3)

julia> one(D)
3×3 Diagonal{Int64,Array{Int64,1}}:         <-- dimensionless
 1  ⋅  ⋅
 ⋅  1  ⋅
 ⋅  ⋅  1
```
should be true.

Reply:
Ah, I interpreted @stevengj 's comment the opposite way. I will update this and add some tests with types that have a dimension.

Reply:
> This would not preserve the underlying container type (as an example: sparsevector).

Who cares?  Why does the underlying container type matter?   An array of ones is not sparse anyway.

Reply:
This way, it would be consistent with `zero`. Not that I think making it a vector is a bad idea though (especially considering the behavior of `zero` and `one` for these matrices backed by ranges).

```
julia> x=sprand(Float64, 10, 10, .1)
10×10 SparseMatrixCSC{Float64,Int64} with 5 stored entries:
  [6 ,  1]  =  0.999066
  [4 ,  4]  =  0.45223
  [3 ,  5]  =  0.187222
  [6 ,  6]  =  0.87619
  [8 ,  9]  =  0.926811

julia> zero(x)
10×10 SparseMatrixCSC{Float64,Int64} with 5 stored entries:
  [6 ,  1]  =  0.0
  [4 ,  4]  =  0.0
  [3 ,  5]  =  0.0
  [6 ,  6]  =  0.0
  [8 ,  9]  =  0.0

julia> one(x)
10×10 SparseMatrixCSC{Float64,Int64} with 10 stored entries:
  [1 ,  1]  =  1.0
  [2 ,  2]  =  1.0
  [3 ,  3]  =  1.0
  [4 ,  4]  =  1.0
  [5 ,  5]  =  1.0
  [6 ,  6]  =  1.0
  [7 ,  7]  =  1.0
  [8 ,  8]  =  1.0
  [9 ,  9]  =  1.0
  [10, 10]  =  1.0

julia> UpperTriangular(x)
10×10 UpperTriangular{Float64,SparseMatrixCSC{Float64,Int64}}:
 0.0  0.0  0.0  0.0      0.0       0.0      0.0  0.0  0.0       0.0
  ⋅   0.0  0.0  0.0      0.0       0.0      0.0  0.0  0.0       0.0
  ⋅    ⋅   0.0  0.0      0.187222  0.0      0.0  0.0  0.0       0.0
  ⋅    ⋅    ⋅   0.45223  0.0       0.0      0.0  0.0  0.0       0.0
  ⋅    ⋅    ⋅    ⋅       0.0       0.0      0.0  0.0  0.0       0.0
  ⋅    ⋅    ⋅    ⋅        ⋅        0.87619  0.0  0.0  0.0       0.0
  ⋅    ⋅    ⋅    ⋅        ⋅         ⋅       0.0  0.0  0.0       0.0
  ⋅    ⋅    ⋅    ⋅        ⋅         ⋅        ⋅   0.0  0.926811  0.0
  ⋅    ⋅    ⋅    ⋅        ⋅         ⋅        ⋅    ⋅   0.0       0.0
  ⋅    ⋅    ⋅    ⋅        ⋅         ⋅        ⋅    ⋅    ⋅        0.0

julia> zero(UpperTriangular(x))
10×10 UpperTriangular{Float64,SparseMatrixCSC{Float64,Int64}}:
 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
  ⋅   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
  ⋅    ⋅   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
  ⋅    ⋅    ⋅   0.0  0.0  0.0  0.0  0.0  0.0  0.0
  ⋅    ⋅    ⋅    ⋅   0.0  0.0  0.0  0.0  0.0  0.0
  ⋅    ⋅    ⋅    ⋅    ⋅   0.0  0.0  0.0  0.0  0.0
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   0.0  0.0  0.0  0.0
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   0.0  0.0  0.0
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   0.0  0.0
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   0.0
```